### PR TITLE
KG builder components user guide

### PR DESCRIPTION
**Pull request by stellasia**
# Description

Includes user guide for the KG builder pipeline with a warning on its experimental nature.

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [x] Documentation update
- [ ] Project configuration change

## Complexity

> **Note**
>
> Please provide an estimated complexity of this PR of either Low, Medium or High
>
>

Complexity: Low

## How Has This Been Tested?
- [ ] Unit tests
- [ ] E2E tests
- [x] Manual tests

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [x] Documentation has been updated
- [ ] Unit tests have been updated
- [ ] E2E tests have been updated
- [ ] Examples have been updated
- [ ] New files have copyright header
- [x] CLA (https://neo4j.com/developer/cla/) has been signed
- [ ] CHANGELOG.md updated if appropriate

### CODE REVIEW

File: docs/source/api.rst [hunk 1]
```diff
@@ -148,25 +149,33 @@ Generation
 **********
 
 LLMInterface
```
Comment (willtai) [inline docs/source/api.rst:151]: I think the convention is the punctuation characters are as long as the text, shall we stick to this? https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#sections

File: docs/source/user_guide_kg_builder.rst [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+
+******************************
+Pipeline structure
+******************************
+
+A Knowledge Graph (KG) construction pipeline requires a few components:
+
+- A document **parser**: extract text from files (PDFs, ...)
+- A document **chunker**: split the text into smaller pieces of text, manageable by the LLM context window (token limit).
+- A chunk **embeder** (optional): compute and store the chunk embeddings
```
Comment (willtai) [inline docs/source/user_guide_kg_builder.rst:25]:
```suggestion
- A chunk **embedder** (optional): compute and store the chunk embeddings
```

File: docs/source/user_guide_kg_builder.rst [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+- A **schema builder**: provide a schema to ground the LLM extracted entities and relations and obtain an easily navigable KG.
+- An **entity and relation extractor**: extract relevant entities and relations from the text.
+- A **Knowledge Graph writer**: write the identified entities and relations to a Neo4j database.
```
Comment (willtai) [inline docs/source/user_guide_kg_builder.rst:28]: Maybe it's just me being very nitpicky, but I think this section is more readable if each sentence doesn't start with "A..." or "An...". I felt the same way about the other user guides

File: docs/source/api.rst (hunk 1 above)
Comment (stellasia) [reply docs/source/api.rst]: Sure, I love conventions :-) Updated in https://github.com/neo4j/neo4j-genai-python/pull/102/commits/4b6590f87b8349daef58d6192a8967ef24dc0ff0

File: docs/source/user_guide_kg_builder.rst [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+**********************************
+
+Below is a list of the different components available in this package, and how to use them.
+
+Each of these components can be run individually:
+
+.. code:: python
+
+    import asyncio
+    # replace `PdfLoader` by a real component name
```
Comment (willtai) [inline docs/source/user_guide_kg_builder.rst:50]: I'm not sure what is meant by "real component name" here

Comment (stellasia) [reply docs/source/user_guide_kg_builder.rst:50]: Will remove it, I used a non existing component at the beginning but found it more useful to choose one real component instead.

File: docs/source/user_guide_kg_builder.rst [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+This package contains the interface and implementations for each of these components, which are detailed in the following sections.
+
+To see an end-to-end example of a Knowledge Graph construction pipeline,
+refer to `this example <https://github.com/neo4j/neo4j-genai-python/blob/main/examples/pipeline/kg_builder.py>`_.
+
+**********************************
+Knowledge Graph Builder Components
+**********************************
+
+Below is a list of the different components available in this package, and how to use them.
```
Comment (willtai) [inline docs/source/user_guide_kg_builder.rst:43]:
```suggestion
Below is a list of the different components available in this package and how to use them.
```

I don't think the comma is required

File: docs/source/user_guide_kg_builder.rst [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+        async def run(self, path: Path) -> PdfDocument:
+            # process file in `path`
+            return PdfDocument(text="text")
+
+
+
+Document Splitter
+=================
+
+Document splitters, as the name indicate, are splitting documents into smaller chunks
```
Comment (willtai) [inline docs/source/user_guide_kg_builder.rst:98]:
```suggestion
Document splitters, as the name indicates, split documents into smaller chunks
```

File: docs/source/user_guide_kg_builder.rst [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+
+    To use OpenAI (embedding or LLM), the `OPENAI_API_KEY` must be in the env vars, for instance using:
+
+    .. code:: python
+
+        import os
+        os.environ["OPENAI_API_KEY"] = "sk-..."
+
+
+If OpenAI is not an option, see :ref:`embedders` to learn how to use sentence-transformer or create your own embedder.
```
Comment (willtai) [inline docs/source/user_guide_kg_builder.rst:162]:
```suggestion
If OpenAI is not an option, see :ref:`embedders` to learn how to use sentence-transformers or create your own embedder.
```

File: docs/source/user_guide_kg_builder.rst [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+    It is not recommended to use it in production yet.
+
+
+******************
+Pipeline structure
+******************
+
+A Knowledge Graph (KG) construction pipeline requires a few components:
+
+- Document **parser**: extract text from files (PDFs, ...)
```
[the remark this answers is not in the document]
Comment (stellasia) [reply docs/source/user_guide_kg_builder.rst:23]: Maybe we should just bold the full expression: **Document parser**, **Chunk embedder**, etc?

File: docs/source/user_guide_kg_builder.rst [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+
+.. warning::
+
+    This feature is still experimental. API changes and bug fixes are expected.
+
+    It is not recommended to use it in production yet.
+
+
+******************
+Pipeline structure
```
Comment (willtai) [inline docs/source/user_guide_kg_builder.rst:18]:
```suggestion
Pipeline Structure
```

File: docs/source/user_guide_kg_builder.rst [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+It can be used in this way:
+
+.. code:: python
+
+    from neo4j_genai.experimental.components.entity_relation_extractor import (
+        LLMEntityRelationExtractor,
+    )
+    from neo4j_genai.llm import OpenAILLM
+
+    extractor = LLMEntityRelationExtractor(
```
Comment (willtai) [inline docs/source/user_guide_kg_builder.rst:243]: I would add a section on inserting a custom prompt / prompt template. I think most users will be very interested in that

Comment (stellasia) [reply docs/source/user_guide_kg_builder.rst:243]: Isn't this covered here: https://github.com/neo4j/neo4j-genai-python/pull/102/files#diff-c37850688512e77f403fe5c0f70c11f569a5ae233777f9859877f2f49dfb4619R306 ?

Comment (willtai) [reply docs/source/user_guide_kg_builder.rst:243]: Sorry I missed that!

Comment (stellasia) [reply docs/source/user_guide_kg_builder.rst:243]: No pb, thanks for the review!

**Pull request merged**